README.md
InputfieldURLChecker module for ProcessWire
================

Admin helper to add button or hotkey to URL type fields to check link in [ProcessWire CMS](http://processwire.com/).

## Features

- enhance FieldtypeURL inputs with a button to open typed url in a new window
- button and/or hotkey modes
- linkify locked URL fields
- force "http://" prefix (optional)
- restrict module to selected templates or fields
- compatible with the default and Reno admin themes
- ajax-loaded tabs and fields ready


## Usage

1. Install the module. All URL type fields will automatically get a URL check button.

1. Go to module settings and tweak options.


## Screens

**In action:**

![InputfieldURLChecker](https://github.com/rolandtoth/InputfieldURLChecker/raw/master/screens/InputfieldURLChecker.gif)

**Options:**

![InputfieldURLChecker options](https://github.com/rolandtoth/InputfieldURLChecker/raw/master/screens/InputfieldURLChecker-options.png)

**Linkify locked field:**

![Linkify locked field](https://github.com/rolandtoth/InputfieldURLChecker/raw/master/screens/InputfieldURLChecker-locked-field.png)


## Options

### Mode

You can choose whether to use a dedicated button or a key/click combo to open URL field contents.

The module currently supports three modes:

- Button: adds a URL check button to the left or to the right side of the field
- Ctrl + Shift + Click: check URL by holding down ctrl+shift and clicking on the field
- Ctrl + Shift + Enter: check URL by holding down ctrl+shift+enter while caret is in the field

### Target options

By default the link is opened in a lightbox. From version 1.01 links can be opened in an iframe that slides in from the left and clicking on its overly closes it.

Note: if the target website doesn't allow embedding from another domain then the iframe mode will fail.

### Force HTTP prefix

If checked, links will always have "http://" prefix, even if the field itself doesn't contain it.

This ensures opening external links instead of relative ones.

### Restrict module

Here you can set templates and/or fields where the module will be enabled. If none selected, module will be active on all templates and all URL type fields.

Note: if used, module will be disabled on all non-listed fields/templates.


## Troubleshooting and feedback

Forum: [https://processwire.com/talk/topic/11209-modulette-inputfieldurlchecker/](https://processwire.com/talk/topic/11209-modulette-inputfieldurlchecker/)


## License

Licensed under the MIT license.

"InputfieldURLChecker" is provided "as-is" without warranty of any kind, express, implied or otherwise, including without limitation, any warranty of merchantability or fitness for a particular purpose. In no event shall the author of this software be held liable for data loss, damages, loss of profits or any other kind of loss while using or misusing this software.


CHANGELOG.md
#Changelog


### 1.1.5 (2016-06-09)

- apply module defaults on install (not only after first save)


### 1.1.4 (2016-04-26)

- allow module on Profile page


### 1.1.3 (2016-01-23)

- restrict module to ProcessPageEdit (thanks to outflux3)


### 1.1.2 (2016-01-06)

- sanitize id (thanks to horst-n)


### 1.1.0 (2015-11-22)

- module was rewritten to use InputfieldURL::render hook instead JavaScript
- fixed locked url fields when field description or notes were present
- locked url fields respect module setting "target" to open url in new window or iframe


### 1.0.5 (2015-11-21)

- fixed disfunctional "Enabled fields" (reported by Macrura)
- extra padding added only to fields that have IUC buttons


TestUrlField.js
/**
 * Test URL Field module for ProcessWire
 * https://goo.gl/GRM0jp
 */

$(document).ready(function () {

    var TUF = {
        parent: '.InputfieldURL',
        selector: 'input:not([type="hidden"])',
        asmSelectPlaceholder: 'data-asm-placeholder',
        lockedFields: '.collapsed7, .collapsed8',
        TUFlink: '<a href="#" class="tuf-link" target="_blank"><i class="fa fa-arrow-right"></i></a>',
        dummyFieldSelector: 'TUF-dummy',
        mode: (config.TUF && config.TUF.mode) ? config.TUF.mode : 'button-left',
        forceHttp: (config.TUF && config.TUF.forceHttp !== 'undefined') ? config.TUF.forceHttp : true,
        enabled_fields: (config.TUF && config.TUF.enabled_fields) ? config.TUF.enabled_fields : false,
        enabled_templates: (config.TUF && config.TUF.enabled_templates) ? config.TUF.enabled_templates : false,
        template: (config.TUF && config.TUF.template) ? config.TUF.template : false
    };


    // stop if template is not among enabled templates
    // but skip if on own module edit page
    if ($('form#ModuleEditForm').attr('action') !== 'edit?name=TestUrlField') {
        if (TUF.enabled_templates && TUF.template) {
            if (TUF.enabled_templates.indexOf(TUF.template) === -1) {
                return true;
            }
        }
    }

    // get button height with a dummy element
    $('body').append('<input class="' + TUF.dummyFieldSelector + '">');
    TUF.btnHeight = $('.' + TUF.dummyFieldSelector).outerHeight() - 2;
    $('.' + TUF.dummyFieldSelector).remove();

    TUF.selector = TUF.parent + ' ' + TUF.selector;

    $(document).on('ready reloaded wiretabclick', initTUF);

    function initTUF() {

        // linkify fields with locked status
        if ($(TUF.parent).is(TUF.lockedFields)) {

            $(TUF.parent).filter(TUF.lockedFields).find('.InputfieldContent').each(function () {

                var content = $(this).html().trim();

                if (content && content !== "" && content !== "&nbsp;" && !$(this).find('a').length) {
                    $(this).wrapInner(function () {
                        return '<a href="' + content + '" target="_blank"></a>';
                    });
                }
            });

            return true;
        }

        else if ($(TUF.selector).length) {

            $(TUF.selector).each(function () {

                var currInput = $(this);

                // continue if field is not among enabled fields
                if (TUF.enabled_fields) {
                    TUF.enabled_fields = TUF.enabled_fields.replace(' ', '');
                    TUF.enabled_fields = '.Inputfield_' + TUF.enabled_fields.replace(',', ', .Inputfield_');

                    if (!$(parent).is(TUF.enabled_fields)) {
                        return true;
                    }
                }

                // set link height
                setTimeout(function () {

                    if (TUF.btnHeight > 0) {
                        currInput.next('.tuf-link').css({
                            'height': TUF.btnHeight + 'px',
                            'line-height': TUF.btnHeight - 1 + 'px'
                        });
                    }
                }, 0);

                // disable adding link and events twice (ajax loaded tabs and fields)
                if (currInput.next('.tuf-link').length) {
                    return true;
                }

                if (TUF.mode === 'button-right' || TUF.mode === 'button-left') {

                    currInput.after(TUF.TUFlink);

                    if (TUF.mode === 'button-left') {
                        $('body').addClass('tuf-left');
                    } else {
                        $('body').addClass('tuf-right');
                    }

                    currInput.on('keyup fieldchange', function () {

                        var url = getUrl(currInput.val(), TUF.forceHttp),
                            link = currInput.next('.tuf-link');

                        link.attr('href', url);

                        if (url) {
                            link.removeClass('tuf-hide');
                        } else {
                            link.addClass('tuf-hide')
                        }
                    });

                    // populate on load
                    currInput.trigger('fieldchange');

                } else if (TUF.mode === 'ctrl-shift-click') {

                    $(TUF.selector).on('click', function (e) {

                        e.preventDefault();

                        if (e.ctrlKey && e.shiftKey) {
                            window.open(getUrl(currInput.val(), TUF.forceHttp));
                        }

                        return false;
                    })

                } else if (TUF.mode === 'ctrl-shift-enter') {

                    $(TUF.selector).on('keydown', function (e) {

                        e.preventDefault();

                        if (e.ctrlKey && e.shiftKey && (e.keyCode == 10 || e.keyCode == 13)) {
                            window.open(getUrl(currInput.val(), TUF.forceHttp));
                        }

                        return false;
                    })
                }
            });
        }
    }


    /**
     * Add placeholder to asmSelect
     * Selector: http://stackoverflow.com/questions/10641258/jquery-select-data-attributes-that-arent-empty#answer-23944081
     */
    $(function () {
        $('select[' + TUF.asmSelectPlaceholder + '!=""][' + TUF.asmSelectPlaceholder + ']').each(function () {

            var placeholder = $(this).data('asmPlaceholder');

            if (placeholder) {
                $(this).parent().find('.asmSelect option:first').attr({
                    'selected': true,
                    'disabled': true
                }).text(placeholder);
            }
        });
    });

});


/**
 * Get url with or without "http://"
 *
 * @param url
 * @param forcePrefix
 * @returns {*}
 */
function getUrl(url, forcePrefix) {
    var prefix = 'http';
    return (forcePrefix && url != "" && url.indexOf(prefix) === -1) ? prefix + '://' + url : url;
}


InputfieldURLChecker.js
/**
 * InputfieldURLChecker module for ProcessWire
 * https://goo.gl/F8CAK8
 */

$(document).ready(function () {

    var IUC = {
        selector: 'a.iuc',
        asmSelectPlaceholder: 'data-asm-placeholder',
        linkHiddenClass: 'iuc-hide',
        lockedClass: 'iuc-locked-link',
        iframeSelector: 'iuc-iframe',
        overlaySelector: 'iuc-overlay',
        dummyFieldSelector: 'IUC-dummy',
        dataMode: 'data-iuc-mode',
        dataTarget: 'data-iuc-target',
        dataForceHttp: 'data-iuc-force-http',
        dataLoaded: 'data-iuc-loaded'
    };

    // get button height with a dummy element
    $('body').append('<input class="' + IUC.dummyFieldSelector + '">');
    IUC.btnHeight = $('.' + IUC.dummyFieldSelector).outerHeight() - 2;
    $('.' + IUC.dummyFieldSelector).remove();

    $(document).on('ready reloaded wiretabclick', initIUC);

    function initIUC() {

        // append empty iframe to the right on start
        if ($('input[' + IUC.dataTarget + '="iframe"]').length && !$('.' + IUC.iframeSelector).length) {
            $('body').append('<iframe class="' + IUC.iframeSelector + '">');
        }

        $(IUC.selector).not('[' + IUC.dataLoaded + '="1"]').each(function () {

            if ($(this).hasClass(IUC.lockedClass)) {
                addLockedFieldMode($(this));
            }

            var currInput = $(this).next('input:not([type="hidden"])');

            // set link height
            setTimeout(function () {
                if (IUC.btnHeight > 0) {
                    currInput.prev(IUC.selector).css({
                        'height': IUC.btnHeight + 'px',
                        //'line-height': IUC.btnHeight - 1 + 'px'
                        'line-height': IUC.btnHeight + 'px'
                    });
                }
            }, 0);

            addButtonMode(currInput);

            $(this).attr(IUC.dataLoaded, 1);
        });


        $('input[' + IUC.dataMode + '!=""][' + IUC.dataMode + ']').not('[' + IUC.dataLoaded + '="1"]').each(function () {

            var mode = $(this).attr(IUC.dataMode);

            if (mode.indexOf('ctrl-shift-click') !== -1) {
                addHotkeyMode($(this), 'click', 'ctrl-shift-click');
            }

            if (mode.indexOf('ctrl-shift-enter') !== -1) {
                addHotkeyMode($(this), 'keydown', 'ctrl-shift-enter');
            }

            $(this).attr(IUC.dataLoaded, 1);
        });
    }


    function addLockedFieldMode(obj) {

        obj.on('click', function (e) {

            var url = $(this).attr('href');

            e.preventDefault();

            return url ? gotoLink(url, obj.attr(IUC.dataTarget)) : false;
        });
    }


    function addButtonMode(obj, side) {

        obj.on('keyup fieldchange', function () {

            var url = getUrl(obj.val(), $(this).attr(IUC.dataForceHttp)),
                link = obj.prev(IUC.selector);

            link.attr('href', url);

            url ? link.removeClass(IUC.linkHiddenClass) : link.addClass(IUC.linkHiddenClass);

        });

        obj.prev(IUC.selector).on('click', function (e) {

            var url = $(this).attr('href'),
                currInput = $(this).next('input:not([type="hidden"])');

            e.preventDefault();

            return url ? gotoLink(url, currInput.attr(IUC.dataTarget)) : false;
        });

        // populate on load
        obj.trigger('fieldchange');

        return false;
    }


    function addHotkeyMode(obj, eventName, mode) {

        obj.on(eventName, function (e) {

            var url = $(this).val(),
                isCtrlShiftPressed = e.ctrlKey && e.shiftKey,
                isEnterPressed = e.keyCode == 10 || e.keyCode == 13;

            if (!url) return true;

            if ((mode === 'ctrl-shift-enter' && isCtrlShiftPressed && isEnterPressed) ||
                (mode === 'ctrl-shift-click' && isCtrlShiftPressed)) {
                gotoLink(getUrl(url, $(this).attr(IUC.dataForceHttp)), $(this).attr(IUC.dataTarget));
                return false;
            }
        });
    }


    /**
     * Open link in a new window/tab or slide in iframe
     * Force new window/tab if document width is below 640px
     *
     * @param url
     * @returns {boolean}
     */
    function gotoLink(url, target) {

        var iframe = $('.' + IUC.iframeSelector);

        if (target === 'new-window' || document.body.scrollWidth < 640) {
            window.open(url);

        } else {

            $('html').css('overflow-y', 'hidden');
            $('body').append('<div class="' + IUC.overlaySelector + '">');
            iframe.addClass('iuc-iframe-animate').attr('src', url);

            iframe.on('load', function () {
                $(this).addClass('iuc-iframe-loaded');
            })
        }

        return false;
    }


    /**
     * Add placeholder to asmSelect
     * Selector: http://stackoverflow.com/questions/10641258/jquery-select-data-attributes-that-arent-empty#answer-23944081
     */
    $(function () {
        $('select[' + IUC.asmSelectPlaceholder + '!=""][' + IUC.asmSelectPlaceholder + ']').each(function () {

            var placeholder = $(this).data('asmPlaceholder');

            if (placeholder) {
                $(this).parent().find('.asmSelect option:first').attr({
                    'selected': true,
                    'disabled': true
                }).text(placeholder);
            }
        });
    });

    // remove overlay on click
    $(document).on('click', '.' + IUC.overlaySelector, function () {
        IUCCloseIframe();
    });

    // remove overlay on pressing ESC
    $(document).bind('keyup', '.' + IUC.overlaySelector, function (e) {
        if (e.keyCode === 27) {
            IUCCloseIframe();
        }
    });

    function IUCCloseIframe() {
        $('.' + IUC.iframeSelector).removeClass('iuc-iframe-animate iuc-iframe-loaded').attr('src', '');
        $('html').css('overflow-y', '');
        $('.' + IUC.overlaySelector).remove();
    }
});


/**
 * Get url with or without "http://"
 *
 * @param url
 * @param forcePrefix
 * @returns {*}
 */
function getUrl(url, forcePrefix) {
    var prefix = 'http';
    return (forcePrefix && url != "" && url.indexOf(prefix) === -1) ? prefix + '://' + url : url;
}
